File: plugins/zebra-label-printer/ios/Plugin/ZebraLabelPrinterPlugin.swift
import Foundation
import Capacitor
import CoreBluetooth

@objc(ZebraLabelPrinterPlugin)
public class ZebraLabelPrinterPlugin: CAPPlugin, CAPBridgedPlugin {
    private func asStringDict(_ value: Any?) -> [String: Any]? {
        switch value {
        case let d as [String: Any]:
            return d
        case let d as NSDictionary:
            return d as? [String: Any]
        default:
            return nil
        }
    }

    public let identifier = "ZebraLabelPrinterPlugin"
    public let jsName = "ZebraLabelPrinter"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPairedPrinters", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setDefaultPrinter", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getDefaultPrinter", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "printLabel", returnType: CAPPluginReturnPromise),
    ]

    /// Entspricht Android: ohne Bluetooth-Nutzungserlaubnis kein sinnvoller Pfad.
    private func bluetoothBlockedCode() -> String? {
        if #available(iOS 13.0, *) {
            switch CBCentralManager.authorization {
            case .denied, .restricted:
                return "BLUETOOTH_UNAUTHORIZED"
            case .notDetermined:
                return "BLUETOOTH_UNAUTHORIZED"
            case .allowedAlways:
                return nil
            @unknown default:
                return "BLUETOOTH_UNAUTHORIZED"
            }
        }
        return nil
    }

    @objc func isAvailable(_ call: CAPPluginCall) {
        if let code = bluetoothBlockedCode() {
            call.resolve([
                "available": false,
                "platform": "ios",
                "reason": code,
            ])
            return
        }
        call.resolve([
            "available": true,
            "platform": "ios",
        ])
    }

    @objc func getPairedPrinters(_ call: CAPPluginCall) {
        if let code = bluetoothBlockedCode() {
            call.resolve(ZebraIosPluginResults.fail(
                code: code,
                message: "Bluetooth-Berechtigung fehlt oder wurde noch nicht erteilt.",
                details: nil
            ))
            return
        }
        // Kein Pendant zu Android getBondedDevices: Geräteliste über Zebra Link-OS / System-API nachrüsten.
        call.resolve(ZebraIosPluginResults.okPrinters([]))
    }

    @objc func setDefaultPrinter(_ call: CAPPluginCall) {
        guard let opts = call.options as? [String: Any] else {
            call.resolve(ZebraIosPluginResults.fail(code: "INVALID_PAYLOAD", message: "Ungültige Aufrufparameter.", details: nil))
            return
        }
        guard opts.keys.contains("target") else {
            call.resolve(ZebraIosPluginResults.fail(code: "INVALID_PAYLOAD", message: "Parameter target fehlt.", details: nil))
            return
        }
        if opts["target"] is NSNull {
            ZebraPrinterStore.saveDefaultPrinterJson(nil)
            call.resolve(ZebraIosPluginResults.okSetDefault())
            return
        }
        guard let target = asStringDict(opts["target"]) else {
            call.resolve(ZebraIosPluginResults.fail(code: "INVALID_PAYLOAD", message: "target ist kein Objekt.", details: nil))
            return
        }
        guard ZebraIosTargetValidation.validateIosTarget(target) else {
            call.resolve(ZebraIosPluginResults.fail(
                code: "INVALID_PAYLOAD",
                message: "Ungültiger PrinterTarget für iOS (kind, native.platform=ios, native.opaque).",
                details: nil
            ))
            return
        }
        guard let json = ZebraIosTargetValidation.targetToJsonString(target) else {
            call.resolve(ZebraIosPluginResults.fail(code: "SDK_ERROR", message: "Druckerdaten konnten nicht serialisiert werden.", details: nil))
            return
        }
        ZebraPrinterStore.saveDefaultPrinterJson(json)
        call.resolve(ZebraIosPluginResults.okSetDefault())
    }

    @objc func getDefaultPrinter(_ call: CAPPluginCall) {
        guard let json = ZebraPrinterStore.loadDefaultPrinterJson(), !json.isEmpty else {
            call.resolve(ZebraIosPluginResults.okDefault(target: nil))
            return
        }
        guard let dict = ZebraIosTargetValidation.jsonStringToTargetDict(json) else {
            call.resolve(ZebraIosPluginResults.fail(code: "SDK_ERROR", message: "Gespeicherte Druckerdaten konnten nicht gelesen werden.", details: nil))
            return
        }
        guard ZebraIosTargetValidation.validateIosTarget(dict) else {
            call.resolve(ZebraIosPluginResults.fail(code: "SDK_ERROR", message: "Gespeicherter Standarddrucker ist ungültig.", details: nil))
            return
        }
        call.resolve(ZebraIosPluginResults.okDefault(target: dict))
    }

    @objc func printLabel(_ call: CAPPluginCall) {
        if let code = bluetoothBlockedCode() {
            call.resolve(ZebraIosPluginResults.fail(
                code: code,
                message: "Bluetooth-Berechtigung fehlt oder wurde noch nicht erteilt.",
                details: nil
            ))
            return
        }
        guard let opts = call.options as? [String: Any] else {
            call.resolve(ZebraIosPluginResults.fail(code: "INVALID_PAYLOAD", message: "Ungültige Aufrufparameter.", details: nil))
            return
        }
        guard asStringDict(opts["payload"]) != nil else {
            call.resolve(ZebraIosPluginResults.fail(code: "INVALID_PAYLOAD", message: "payload fehlt.", details: nil))
            return
        }
        guard let zpl = opts["zpl"] as? String, !zpl.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            call.resolve(ZebraIosPluginResults.fail(
                code: "INVALID_PAYLOAD",
                message: "zpl fehlt oder ist leer – in JS mit buildZplFromPayloadV1 erzeugen.",
                details: nil
            ))
            return
        }

        var timeoutMs: Int64 = 15_000
        var targetDict: [String: Any]?
        if let printOptions = asStringDict(opts["printOptions"]) {
            if let n = printOptions["timeoutMs"] as? NSNumber {
                timeoutMs = n.int64Value
            } else if let n = printOptions["timeoutMs"] as? Int64 {
                timeoutMs = n
            } else if let n = printOptions["timeoutMs"] as? Int {
                timeoutMs = Int64(n)
            }
            targetDict = asStringDict(printOptions["target"])
        }
        if targetDict == nil, let json = ZebraPrinterStore.loadDefaultPrinterJson(), !json.isEmpty {
            targetDict = ZebraIosTargetValidation.jsonStringToTargetDict(json)
        }
        guard let resolvedTarget = targetDict else {
            call.resolve(ZebraIosPluginResults.fail(
                code: "NO_DEFAULT_PRINTER",
                message: "Kein Drucker gewählt – setDefaultPrinter oder printOptions.target setzen.",
                details: nil
            ))
            return
        }
        guard ZebraIosTargetValidation.validateIosTarget(resolvedTarget) else {
            call.resolve(ZebraIosPluginResults.fail(
                code: "PLATFORM_MISMATCH",
                message: "Druckerziel ist kein gültiges iOS-Ziel.",
                details: nil
            ))
            return
        }
        guard let native = resolvedTarget["native"] as? [String: Any],
              let opaque = native["opaque"] as? String else {
            call.resolve(ZebraIosPluginResults.fail(code: "INVALID_PAYLOAD", message: "native.opaque fehlt.", details: nil))
            return
        }

        switch ZebraIosTransport.sendZplUtf8(opaque: opaque, zpl: zpl, timeoutMs: timeoutMs) {
        case let .success(bytesSent, durationMs):
            call.resolve(ZebraIosPluginResults.okPrint(bytesSent: bytesSent, durationMs: durationMs))
        case let .failure(code, message, details):
            call.resolve(ZebraIosPluginResults.fail(code: code, message: message, details: details))
        }
    }
}

File: plugins/zebra-label-printer/android/src/main/java/de/vico/app/zebralabelprinter/ZebraLabelPrinterPlugin.java
package de.vico.app.zebralabelprinter;

import android.Manifest;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothManager;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.Set;
import java.util.regex.Pattern;
import org.json.JSONException;

@CapacitorPlugin(name = "ZebraLabelPrinter")
public class ZebraLabelPrinterPlugin extends Plugin {

    private static final Pattern MAC_PATTERN = Pattern.compile("^([0-9A-Fa-f]{2}:){5}[0-9A-Fa-f]{2}$");

    private boolean hasBluetoothPermission(final Context ctx) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            return ContextCompat.checkSelfPermission(ctx, Manifest.permission.BLUETOOTH_CONNECT) == PackageManager.PERMISSION_GRANTED;
        }
        return ContextCompat.checkSelfPermission(ctx, Manifest.permission.BLUETOOTH) == PackageManager.PERMISSION_GRANTED;
    }

    private static boolean validateAndroidTarget(final JSObject t) {
        if (t == null) {
            return false;
        }
        if (!"zebra_zq220".equals(t.optString("kind", ""))) {
            return false;
        }
        final JSObject n = t.getJSObject("native");
        if (n == null) {
            return false;
        }
        if (!"android".equals(n.optString("platform", ""))) {
            return false;
        }
        final String mac = n.optString("opaque", "");
        return MAC_PATTERN.matcher(mac).matches();
    }

    @Nullable
    private static JSObject deviceToTarget(final BluetoothDevice d) {
        if (d == null) {
            return null;
        }
        final String mac = d.getAddress();
        final String name = d.getName();
        final JSObject nativeObj = new JSObject();
        nativeObj.put("platform", "android");
        nativeObj.put("opaque", mac);
        final JSObject t = new JSObject();
        t.put("kind", "zebra_zq220");
        t.put("id", mac);
        final String label = name != null && !name.isEmpty() ? name : "Bluetooth " + mac.substring(Math.max(0, mac.length() - 5));
        t.put("displayName", label);
        final String clean = mac.replace(":", "");
        if (clean.length() >= 4) {
            t.put("lastFour", clean.substring(clean.length() - 4));
        }
        t.put("native", nativeObj);
        return t;
    }

    @PluginMethod
    public void isAvailable(final PluginCall call) {
        final Context ctx = getContext();
        final JSObject r = new JSObject();
        r.put("platform", "android");
        if (!ZebraBluetoothSppSender.hasBluetoothAdapter(ctx)) {
            r.put("available", false);
            r.put("reason", "UNSUPPORTED_OS_VERSION");
            call.resolve(r);
            return;
        }
        if (!hasBluetoothPermission(ctx)) {
            r.put("available", false);
            r.put("reason", "BLUETOOTH_UNAUTHORIZED");
            call.resolve(r);
            return;
        }
        if (!ZebraBluetoothSppSender.isBluetoothEnabled(ctx)) {
            r.put("available", false);
            r.put("reason", "BLUETOOTH_OFF");
            call.resolve(r);
            return;
        }
        r.put("available", true);
        call.resolve(r);
    }

    @PluginMethod
    public void getPairedPrinters(final PluginCall call) {
        final Context ctx = getContext();
        if (!hasBluetoothPermission(ctx)) {
            ZebraPluginResults.resolveFail(call, "BLUETOOTH_UNAUTHORIZED", "Berechtigung für Bluetooth fehlt (z. B. BLUETOOTH_CONNECT).", null);
            return;
        }
        if (!ZebraBluetoothSppSender.isBluetoothEnabled(ctx)) {
            ZebraPluginResults.resolveFail(call, "BLUETOOTH_OFF", "Bluetooth ist ausgeschaltet.", null);
            return;
        }
        final BluetoothManager bm = (BluetoothManager) ctx.getSystemService(Context.BLUETOOTH_SERVICE);
        if (bm == null) {
            ZebraPluginResults.resolveFail(call, "SDK_ERROR", "BluetoothManager nicht verfügbar.", null);
            return;
        }
        final BluetoothAdapter adapter = bm.getAdapter();
        if (adapter == null) {
            ZebraPluginResults.resolveFail(call, "SDK_ERROR", "BluetoothAdapter nicht verfügbar.", null);
            return;
        }
        final Set<BluetoothDevice> bonded = adapter.getBondedDevices();
        final JSArray arr = new JSArray();
        if (bonded != null) {
            try {
                for (BluetoothDevice d : bonded) {
                    final JSObject t = deviceToTarget(d);
                    if (t != null) {
                        arr.put(t);
                    }
                }
            } catch (JSONException e) {
                ZebraPluginResults.resolveFail(call, "SDK_ERROR", "Geräteliste konnte nicht erstellt werden.", e.getMessage());
                return;
            }
        }
        call.resolve(ZebraPluginResults.okPrinters(arr));
    }

    @PluginMethod
    public void setDefaultPrinter(final PluginCall call) {
        final Context ctx = getContext();
        final JSObject data = call.getData();
        if (!data.has("target")) {
            ZebraPluginResults.resolveFail(call, "INVALID_PAYLOAD", "Parameter target fehlt.", null);
            return;
        }
        if (data.isNull("target")) {
            ZebraPrinterPrefs.saveDefaultPrinter(ctx, null);
            call.resolve(ZebraPluginResults.okSetDefault());
            return;
        }
        final JSObject target = call.getObject("target");
        if (!validateAndroidTarget(target)) {
            ZebraPluginResults.resolveFail(call, "INVALID_PAYLOAD", "Ungültiger PrinterTarget für Android (kind/native.opaque).", null);
            return;
        }
        ZebraPrinterPrefs.saveDefaultPrinter(ctx, target.toString());
        call.resolve(ZebraPluginResults.okSetDefault());
    }

    @PluginMethod
    public void getDefaultPrinter(final PluginCall call) {
        final Context ctx = getContext();
        final String json = ZebraPrinterPrefs.loadDefaultPrinterJson(ctx);
        if (json == null || json.isEmpty()) {
            call.resolve(ZebraPluginResults.okDefault(null));
            return;
        }
        try {
            final JSObject t = new JSObject(json);
            if (!validateAndroidTarget(t)) {
                ZebraPluginResults.resolveFail(call, "SDK_ERROR", "Gespeicherter Standarddrucker ist ungültig.", null);
                return;
            }
            call.resolve(ZebraPluginResults.okDefault(t));
        } catch (Exception e) {
            ZebraPluginResults.resolveFail(call, "SDK_ERROR", "Gespeicherte Druckerdaten konnten nicht gelesen werden.", e.getMessage());
        }
    }

    @PluginMethod
    public void printLabel(final PluginCall call) {
        final Context ctx = getContext();
        if (!hasBluetoothPermission(ctx)) {
            ZebraPluginResults.resolveFail(call, "BLUETOOTH_UNAUTHORIZED", "Berechtigung für Bluetooth fehlt.", null);
            return;
        }
        if (!ZebraBluetoothSppSender.isBluetoothEnabled(ctx)) {
            ZebraPluginResults.resolveFail(call, "BLUETOOTH_OFF", "Bluetooth ist ausgeschaltet.", null);
            return;
        }
        final JSObject payload = call.getObject("payload");
        if (payload == null) {
            ZebraPluginResults.resolveFail(call, "INVALID_PAYLOAD", "payload fehlt.", null);
            return;
        }
        final String zpl = call.getString("zpl");
        if (zpl == null || zpl.trim().isEmpty()) {
            ZebraPluginResults.resolveFail(
                call,
                "INVALID_PAYLOAD",
                "zpl fehlt oder ist leer – in JS mit buildZplFromPayloadV1(payload) erzeugen und mitsenden.",
                null
            );
            return;
        }

        final JSObject printOptions = call.getObject("printOptions");
        long timeoutMs = 15000;
        JSObject targetObj = null;
        if (printOptions != null) {
            timeoutMs = printOptions.optLong("timeoutMs", 15000);
            if (printOptions.has("target") && !printOptions.isNull("target")) {
                targetObj = printOptions.getObject("target");
            }
        }
        if (targetObj == null) {
            final String saved = ZebraPrinterPrefs.loadDefaultPrinterJson(ctx);
            if (saved != null && !saved.isEmpty()) {
                try {
                    targetObj = new JSObject(saved);
                } catch (Exception ignored) {}
            }
        }
        if (targetObj == null) {
            ZebraPluginResults.resolveFail(call, "NO_DEFAULT_PRINTER", "Kein Drucker gewählt – setDefaultPrinter oder printOptions.target setzen.", null);
            return;
        }
        if (!validateAndroidTarget(targetObj)) {
            ZebraPluginResults.resolveFail(call, "PLATFORM_MISMATCH", "Druckerziel ist kein gültiges Android-Ziel.", null);
            return;
        }
        final JSObject nativeObj = targetObj.getJSObject("native");
        final String mac = nativeObj != null ? nativeObj.optString("opaque", "") : "";
        final ZebraBluetoothSppSender.SendOutcome out = ZebraBluetoothSppSender.sendZplUtf8(ctx, mac, zpl, timeoutMs);
        if (!out.success) {
            ZebraPluginResults.resolveFail(call, out.errorCode, out.message, out.details);
            return;
        }
        call.resolve(ZebraPluginResults.okPrint(out.bytesSent, out.durationMs));
    }
}
